src/main/java/org/opensearch/ad/task/ADTaskManager.java

```diff
@@ -1020,49 +1019,101 @@ public void stopDetector(
         }));
     }
 
-    private void resetLatestHistoricalDetectorTaskState(
+    private <T> void resetLatestHistoricalDetectorTaskState(
         List<ADTask> adTasks,
         Consumer<List<ADTask>> function,
-        TransportService transportService
+        TransportService transportService,
+        ActionListener<T> listener
     ) {
-        List<ADTask> longRunningHistoricalTasks = adTasks
-            .stream()
-            .filter(t -> t.isHistoricalTask() && !t.isEntityTask() && !isADTaskEnded(t) && lastUpdateTimeExpired(t))
-            .collect(Collectors.toList());
-
-        if (longRunningHistoricalTasks.size() > 0) {
-            ADTask adTask = longRunningHistoricalTasks.get(0);
-            resetHistoricalDetectorTaskState(adTask, () -> function.accept(adTasks), transportService);
-        } else {
-            function.accept(adTasks);
+        List<ADTask> runningHistoricalTasks = new ArrayList<>();
+        List<ADTask> runningRealtimeTasks = new ArrayList<>();
+        for (ADTask adTask : adTasks) {
+            if (!adTask.isEntityTask() && !adTask.isDone()) {
+                if (!adTask.isHistoricalTask()) {
+                    // try to reset task state if realtime task is not ended
+                    runningRealtimeTasks.add(adTask);
+                } else {
+                    // try to reset task state if historical task not updated for 2 piece intervals
+                    runningHistoricalTasks.add(adTask);
+                }
+            }
         }
+
+        resetHistoricalDetectorTaskState(
+            runningHistoricalTasks,
+            () -> resetRealtimeDetectorTaskState(runningRealtimeTasks, () -> function.accept(adTasks), transportService, listener),
+            transportService,
+            listener
+        );
     }
 
-    private void resetHistoricalDetectorTaskState(ADTask adTask, AnomalyDetectorFunction function, TransportService transportService) {
+    private <T> void resetRealtimeDetectorTaskState(
+        List<ADTask> runningRealtimeTasks,
+        AnomalyDetectorFunction function,
+        TransportService transportService,
+        ActionListener<T> listener
+    ) {
+        if (isNullOrEmpty(runningRealtimeTasks)) {
+            function.execute();
```

Comment:
Does function guarantee to return using listener?

Reply:
Yes, the listener will return in code block wrapped in function.

Reply:
Checked the call graph, this function will be called by ADTaskManager.getAndExecuteOnLatestADTask.  Since getAndExecuteOnLatestADTask is public, users can call it anyway they want and may miss the listener return.

I checked a few callers of ADTaskManager.getAndExecuteOnLatestADTask.  Does https://github.com/ylwu-amzn/anomaly-detection-2/blob/reset_realtime_task_state/src/main/java/org/opensearch/ad/rest/handler/IndexAnomalyDetectorActionHandler.java#L257 return listener?

Reply:
listener is one `field` of class `IndexAnomalyDetectorActionHandler`, I see listener returned.

Reply:
I see.  Can you add comments to remind the caller to return listener?

Reply:
sure, added